### PR TITLE
Grackle Tests

### PR DESCRIPTION
**Pull request by mabruzzo**
**Overview**
This PR introduces 2 automated tests for `EnzoMethodGrackle`. These are the first tests to check that Grackle-related results remain consistent across commits. In both of these tests, the code sets up parallel one-zone models without hydrodynamics. More details are provided about the test below:
1. `method_grackle_general`: This uses the `null` method to set the simulation timestep (each timestep extends over multiple heating/cooling times). After a fixed amount of simulation time, the data is written to disk and the new `tools/field_summary.py` script is used to compute several summary statistics for relevant fields (`"pressure"`, `"temperature"`, `"cooling_time"`) and compare them against previously computed reference values. The summary statistics include the mean, standard deviation, min, min-location, max, and max-location.
2.  `method_grackle_cooling_dt`: This uses a similar setup to the first test, except that the timesteps are set to be a fixed fraction of the minimum magnitude of the heating/cooling time. This test checks the final simulation time, after a fixed number of cycles, against some reference value.

An advantage of not setting the timestep in the first test to be a fixed fraction of the minimum heating/cooling timescale is that it increases the chance that we might witness race-conditions (if any exist or get introduced) when the simulation is run in SMP-mode.

**More Details**

Unfortunately these tests may be a little fragile given that Grackle is an external dependency and incremental changes to Grackle may slightly change results. Hopefully, we can adjust the tolerances to be flexible enough that this isn't a major problem (but the tests remain meaningful). In any case, I think these tests are definitely an improvement over having nothing. 

The enzo-dev style of regression/answer testing would certainly be a superior approach (it would facilitate comparisons between two revisions of Enzo-E on the same machine using a single build of Grackle). However, the machinery behind that testing system remains opaque to me. It would be great if we could move to that system sooner rather than later. I expect that my creation of `tools/field_summary.py` was reinventing the wheel to some degree, and it would be great to rely on existing machinery. As an aside, `tools/field_summary.py` could probably be reused in the short-term for the creation of other tests (e.g. the 1D Zel'dovich pancake tests mentioned in Issue #123)

An obstacle to creating these tests was that the path to the Grackle data files can't be hard-coded into the input files. To get around this issue, I have introduced  `tools/gen_grackle_testing_file.py` which writes a new input file to disk that simply has an include-directive (i.e. to include the contents of `input/method_grackle_general.in` or `input/method_grackle_cooling_dt.in` and overwrites the value of the `Method:grackle:data_file` parameter. I choose to make this functionality and `tools/field_summary.py` into a command-line tool because of my uncertainty over the future format of the testing system (giving the pending transition of the build-system from Scons to CMake).

While implementing `test/MethodGrackle/SConscript` I broke parts of the test into individual build steps. I realize that may be somewhat difficult to follow. Let me know if you think I should refactor running the simulation and executing `tools/field_summary.py` into a single build step for the `method_grackle_general` test.

**Removing `-ffast-math` from default build**
This PR was delayed by issues I encountered with reproducibility. The tests ran fine on different machines with the same compiler version (e.g. on my personal laptop and the CI infrastructure). However, we got surprisingly large differences (the biggest relative difference was maybe ~0.1% for double-precision tests) when we ran the test using a binary with a different compiler version (e.g. my work machine). In particular I had encountered issues with versions 7.5 and 9.4 of the gcc compiler.

It turns out that these issues arose because of the value-unsafe floating-point optimizations enabled by the `-ffast-math` flag (which was enabled by default on `linux-gnu` builds). It took me a while to figure this out since Grackle itself is never compiled with this flag. As far as I can tell, most of the differences creep in during initialization of fields in `EnzoInitialGrackleTest`.

As a consequence of this issue, I have removed the `-ffast-math` flag from the default build configuration. I don't think this should be controversial since we were already moving towards this change in PR #139. Removing this flag required the following modifications to existing tests:
1. the expected values for the VL+CT's dual_energy_cloud test needed to be updated (there was already a conversation about needing this change in PR #139).
2. the final time for the PPML tests had to be updated for binaries compiled with single precision
3. I needed to change the precision for the `mass` constant in the cosmology test from `"double"` to `"default"`. Without this change, the test crashed for binaries compiled with single precision (without the `-ffast-math`). This problem was discussed in detail in Issue #154.

While setting the tolerances for the double-precision version of the `method_grackle_general` test, I took some care to make sure they were loose enough to avoid issues when running the test on my 2 machines with different compiler versions. I didn't take as much care for the single-precision tests (I figure we can loosen the tolerances on those tests as it becomes necessary)

### CODE REVIEW

**Review by brittonsmith — APPROVED** [review]
This looks good to me. I had one very minor comment and a somewhat unrelated question.

File: input/Grackle/grackle.incl
Comment (mabruzzo) [reply input/Grackle/grackle.incl:44]: The automatic field creation is able to handle all of Grackle's species fields. I can definitely remove these from the input file.

File: input/Grackle/grackle.incl
Comment (mabruzzo) [reply input/Grackle/grackle.incl:73]: The automatic field creation is already smart enough to add all of the species fields (and `"metal_density"` field) to the `color` group when Grackle is in use. (If Grackle is not being used, the fields will not be added to the `color` group - even if they are already defined. That may be worth revisiting in the future).

At this time, you need to manually label `"derived"` fields (like `"temperature"` or `"pressure"`).

File: input/Grackle/grackle.incl
Comment (gregbryan) [inline input/Grackle/grackle.incl:27]:
```suggestion
    length  = 3.086E18;       # 1 pc in cm - does not actually matter
```

File: doc/source/tests/grackle.rst
Comment (gregbryan) [inline doc/source/tests/grackle.rst:17]: Can you explain this a bit more -- why does it test SMP mode?

**Review by gregbryan — APPROVED** [review]
Looks good to me (one minor correction and one minor docs request).

**Comment by peeples** [comment]
@mabruzzo Please merge this once ready; looks like very little left to do!

**Comment by mabruzzo** [comment]
@peeples, I had to introduce 2 new scripts to get this to work properly. I'm not sure whether @gregbryan or @brittonsmith would want to look at this again.

I'm inclined to pull the trigger, since these new scripts effectively transcribe the logic that I had previously encoded in an SConscript file (and the logic should now be more readable). Plus, merging this in might help with wrapping up the multispecies cosmology PR (#143). But, I'll defer to your judgement.

**Comment by peeples** [comment]
Let's merge, and if more issues arise, they can be fixed in a small-and-fast PR. Thanks!

**Pull request merged**
